File: ruby/iforevents-segment/test/segment_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "iforevents-segment"

class SegmentIntegrationTest < Minitest::Test
  class FakeClient
    attr_reader :calls

    def initialize
      @calls = []
    end

    %i[identify track page].each { |m| define_method(m) { |attrs| @calls << [m, attrs] } }

    def flush
      @calls << [:flush]
    end
  end

  def test_forwarding
    fake = FakeClient.new
    client = Iforevents::Client.new([Iforevents::Integrations::Segment.new(client: fake)], context: -> { {} })
    client.init
    client.track("anon")
    client.identify("u", plan: "pro")
    client.track("paid", amount: 1)
    client.page("Home", to_route: "/")
    client.reset
    client.track("again")
    client.shutdown
    assert_equal [:track, { anonymous_id: "server", event: "anon", properties: {}, timestamp: fake.calls[0][1][:timestamp] }], fake.calls[0]
    assert_equal [:identify, { user_id: "u", traits: { "plan" => "pro" } }], fake.calls[1]
    assert_equal "u", fake.calls[2][1][:user_id]
    assert_equal({ "plan" => "pro", "amount" => 1 }, fake.calls[2][1][:properties])
    assert_equal [:page, "Home", "/"], [fake.calls[3][0], fake.calls[3][1][:name], fake.calls[3][1][:properties]["to_route"]]
    assert_equal "server", fake.calls[4][1][:anonymous_id]
    assert_equal [:flush], fake.calls.last
  end

  def test_real_client_constructs
    integration = Iforevents::Integrations::Segment.new(write_key: "wk", client_options: { stub: true })
    integration.init
  end
end

File: go/integrations/segment/segment_test.go
package segment

import (
	"context"
	"testing"

	iforevents "github.com/innovafour/iforevents-sdks/go"
	analytics "github.com/segmentio/analytics-go/v3"
)

type fake struct {
	msgs   []analytics.Message
	closed bool
}

func (f *fake) Enqueue(m analytics.Message) error { f.msgs = append(f.msgs, m); return nil }
func (f *fake) Close() error                      { f.closed = true; return nil }

func TestForwarding(t *testing.T) {
	ctx := context.Background()
	f := &fake{}
	c := iforevents.New(iforevents.WithIntegrations(New(Config{Client: f})), iforevents.WithContext(func() iforevents.Properties { return nil }))
	c.Init(ctx)
	c.Track(ctx, "anon", nil)
	c.Identify(ctx, "u", iforevents.Properties{"plan": "pro"})
	c.Track(ctx, "paid", iforevents.Properties{"amount": 1})
	c.Page(ctx, "Home", nil, iforevents.PageOptions{ToRoute: "/"})
	c.Reset(ctx)
	c.Track(ctx, "again", nil)
	c.Shutdown(ctx)
	if len(f.msgs) != 5 || !f.closed {
		t.Fatalf("msgs %d closed %v", len(f.msgs), f.closed)
	}
	if m := f.msgs[0].(analytics.Track); m.AnonymousId != "server" || m.UserId != "" {
		t.Fatalf("anon: %+v", m)
	}
	if m := f.msgs[1].(analytics.Identify); m.UserId != "u" || m.Traits["plan"] != "pro" {
		t.Fatalf("identify: %+v", m)
	}
	if m := f.msgs[2].(analytics.Track); m.UserId != "u" || m.Properties["amount"] != 1 || m.Properties["plan"] != nil || m.Timestamp.IsZero() {
		t.Fatalf("paid: %+v", m)
	}
	if m := f.msgs[3].(analytics.Page); m.Name != "Home" || m.Properties["to_route"] != "/" {
		t.Fatalf("page: %+v", m)
	}
	if m := f.msgs[4].(analytics.Track); m.AnonymousId != "server" {
		t.Fatalf("after reset: %+v", m)
	}
}

func TestRealClientCreated(t *testing.T) {
	i := New(Config{WriteKey: "wk", Configure: func(c *analytics.Config) { c.BatchSize = 1000 }})
	if err := i.Init(context.Background()); err != nil {
		t.Fatal(err)
	}
	if err := i.Shutdown(context.Background()); err != nil {
		t.Fatal(err)
	}
}
